### ISSUE TITLE
LUT for 32-bit int

### ISSUE THREAD

**Issue by Daniel-Alievsky** [opencv/opencv#13470]
Table translation (LUT) is a frequently function while image processing and analysis. But it is usually necessary not for byte (CV_8U), but for 32-bit integer matrices. For example, we have a map of 100000 objects, recognized on the image (like connected components or results of superpixel), and we join them to some 1000 clusters according different rules (on the base of color, borders between objects, gradients, etc.) - after this, we need to remap the matrix by table int[100000].

Of course, we always can write the loop manually. But, on the one hand, OpenCV could provide maximally quick implementation, optimized for GPU, multithreading and other cases. On the other hand, many algorithms are written not in C++, but in Python, Java (like in my case), etc. Simple loop in Java works essentially slower, than your current LUT implementation, and it is especially important when all data are stored in GPU memory (UMat).

I think it could be a good idea to add LUT version for CV_32S input type in next OpenCV version. What do you think?

**Comment by kinchungwong** [comment]
On one hand, table translation is a useful function to have. For example, converting a connected-components labeled result (or any other segmentation output) into RGB colored image can be done with table translation.

(Please be reminded that the LUT function shouldn't be too limited with respect to data types. It should support multi-channel types as well. For example, my example above requires 32SC1 to 8UC3.)

On the other hand, from a C++ programmer's point of view, there is no need for OpenCV to implement it, because a C++ programmer can write one trivially, and there is no such thing as magical performance advantage that OpenCV can offer. (Multi-threading helps; loop-unrolling helps; but there is no performance gain from using SIMD or scatter-gather instructions.) From Java programmer's point of view, it depends on the particular JIT's willingness to optimize such loop, when the loop is written in Java. Be sure to check the JIT's documentation.

If your project already uses a combination of Java and C++ (such as via JNI or FFI), it might be easier to just ask the C++ team to add a few functions, rather than waiting for OpenCV to provide this function, because it may take more than a year for OpenCV to go from 4.0.1 to 5.0. (The "nightly" code base evolves much faster than releases, but it might require more knowledge to know how to build it and use it with Java.)

**Comment by Daniel-Alievsky** [comment]
Of course, Java can provide high speed, comparable with simple C loop, especially with multithreading. And I really have a Java function for table translation. 
Problems appear when we want to use OpenCV functions. It is a good idea to build a solution from "pure" OpenCV functions, processing UMat. Then we can hope (theoretically) that all image processing will be performed in GPU memory, without copying data into usual RAM. Such approach allows to build very quick algorithms, where each operation requires <1ms for processing a frame 2048x2048. OpenCV here works as a quick "black box", and we can hope that every new OpenCV version will work faster.
But mixing OpenCV code with Java code, of course, slows down the performance. We must copy data to Java array, process, copy back... It is inconvenient and inefficient.
It seems LUT translation for 32-bit integer should be a good improvement for the next OpenCV version.
